devito/types/basic.py

```diff
@@ -664,6 +664,15 @@ def __init_finalize__(self, *args, **kwargs):
 
     __hash__ = Cached.__hash__
 
+    def subs(self, *args, **kwargs):
+        on_grid = kwargs.pop('on_grid', self.on_grid)
+        newobj = super(AbstractFunction, self).subs(*args, **kwargs)
+        try:
+            newobj._on_grid = on_grid
```

Comment:
changing state, not acceptable

also, don't understand how can you have an AttributeError here, since you're setting an attribute

Reply:
- Not changing state, this is creating a new object and initializing it, not changing an existing one
- Because `Dummy` ends up here too in sympy.solve for example that thros an AttributeError.

Reply:
this is changing state

if you don't want it to be changing state, it must occur through `__init__` or `__new__`